Add Bayes peak probability

ahiguera-mx opened the pull request:
_Before you submit this PR: make sure to put all operations-related information in a wiki-note, a PR should be about code and is publicly accessible_
Add Bayes peak probability
## What does the code in this PR do / what does it improve?
Bayes peak probability computation
## Can you briefly describe how it works?
Bayes peak probability computation and propagate this information up to event level 
## Can you give a minimal working example (or illustrate with a figure)?

_Please include the following if applicable:_
  - [ ] _Update the docstring(s)_
  - [ ] _Update the documentation_
  - [ ] _Tests to check the (new) code is working as desired._
  - [ ] _Does it solve one of the open issues on github?_

### _Notes on testing_
 - _Until the automated tests pass, please mark the PR as a draft._
 - _On the XENONnT fork we test with database access, on private forks there is no database access for security considerations._

All _italic_ comments can be removed from this template.

WenzDaniel reviewed (commented):
Here are already a few comments. It would be helpful if you could break down your monolithic computes code into a few smaller functions. This will increase readability and makes the code more transparent. In addition I think it would be good to get some unittests for these smaller functions.

ahiguera-mx commented on straxen/plugins/bayes_peaklet_classification.py:
```diff
@@ -0,0 +1,132 @@
+    Bayes Peaklet classification
+    """
+    provides = 'peaklet_classification_bayes'
+    depends_on = ('peaklets',)
```
This would imply retraining, and for most of the cases we focus on s1_prob

JoranAngevaare replied:
Ah I was about to make the same suggestion as Daniel. There is one extra benefit and that would be that it would be much faster since we are running on much fewer Sx since the S2s are split over many peaklets by design

ahiguera-mx replied:
Right, since we are doing classification we decided to switch to peak as you guys suggested

ahiguera-mx commented on straxen/plugins/bayes_peaklet_classification.py:
```diff
@@ -0,0 +1,132 @@
+
+    def setup(self):
+
+        self.class_prior = np.array([1. / self.classes for j in range(self.classes)])
```
fixed

ahiguera-mx commented on straxen/plugins/bayes_peaklet_classification.py:
```diff
@@ -0,0 +1,132 @@
+        default=np.log(0.5), #best-fit value from optimization of ROC curve
+        help='S2 log prob value threshold, above this value type=2'
+    )
+    num_nodes = straxen.URLConfig(
+        default=50,
+        help='Number of nodes'
+    )
+    classes = straxen.URLConfig(
+        default=2,
+        help='Number of classes'
```
fixed

ahiguera-mx commented on straxen/plugins/bayes_peaklet_classification.py:
```diff
@@ -0,0 +1,132 @@
+        bayes_ptype = class_assignments
+        '''
+
+        return dict(time=peaklets['time'],
+                    dt=peaklets['dt'],
+                    channel=-1,
+                    length=peaklets['length'],
+                    s1_ln_prob=s1_ln_prob,
+                    s2_ln_prob=s2_ln_prob
+                    )
```
we are not doing class assignments in the plugin in, good catch, fixed

ahiguera-mx commented on straxen/plugins/bayes_peaklet_classification.py:
```diff
@@ -0,0 +1,132 @@
+
+        s1_ln_prob = lnposterior_normed[:, 0]
+        s2_ln_prob = lnposterior_normed[:, 1]
+        '''
+        # Probabilities to classes.
+        # If doing peak classification, assign classes
+        # At the moment we are only looking at the posterior probability 
+        
+        class_assignments = np.zeros(len(peaklets))
+        C_S1 = s2_ln_prob < self.s2_prob_threshold
+        C_S2 = s2_ln_prob >= self.s2_prob_threshold
+
+        class_assignments[C_S1] = 1
+        class_assignments[C_S2] = 2
+        bayes_ptype = np.zeros(len(peaklets), dtype=np.int8)
+        bayes_ptype = class_assignments
+        '''
```
fixed

ahiguera-mx commented:
@WenzDaniel @sophiafarrell 
Thanks for the very useful comments, I've implemented all of them. More importantly, I've added unit test for the plugin. 
Hopefully this can go into the next release so analysts can start use it. 
As you may notices from the pytest this PR is from a fork... yes I should have done a branch from master... sorry about that

JoranAngevaare commented on straxen/plugins/bayes_peaklet_classification.py:
```diff
@@ -0,0 +1,141 @@
+        'conditional_probabilities_and_bins_v1_w_global_v6.npz?fmt=npy',
+        help='Bayes configuration file, conditional probabilities tables and Bayes discrete bins'
+    )
+    num_nodes = straxen.URLConfig(
+        default=50,
+        help='Number of attributes(features) per waveform and quantile'
+    )
+    classes = straxen.URLConfig(
+        default=2,
+        help='Number of label classes S1(1)/S2(2)'
+    )
```
Maybe keep in mind that these config names like `'classes'` are visible to all other plugins, as such, I would advice making it a bit more explicit: `classes->n_bayes_classes / bayes_peak_types`, `bayes_nodes` or something like it?

ahiguera-mx replied:
good point, fixed

JoranAngevaare commented on straxen/plugins/bayes_peaklet_classification.py:
```diff
@@ -0,0 +1,141 @@
+    provides = 'peaklet_classification_bayes'
+    depends_on = ('peaklets',)
+    __version__ = '0.0.1'
+    dtype = (strax.peak_interval_dtype
```
maybe the strax.time_fields is more natural here?

ahiguera-mx replied:
fixed

JoranAngevaare commented on straxen/plugins/bayes_peaklet_classification.py:
```diff
@@ -0,0 +1,141 @@
+    data = peaklets['data'].copy()
+    dts = peaklets['dt'].copy()
+    data[data < 0.0] = 0.0
+    for i, p in enumerate(peaklets):
+        sample_number = np.arange(0, num_samples+1, 1)*p['dt']
+        frac_of_cumsum = np.append([0.0], np.cumsum(data[i, :]) / np.sum(data[i, :]))
+        cumsum_steps = np.interp(np.linspace(0., 1., num_nodes, endpoint=False), frac_of_cumsum, sample_number)
+        cumsum_steps = np.append(cumsum_steps, sample_number[-1])
+        quantiles[i, :] = cumsum_steps[1:] - cumsum_steps[:-1]
```
This looks like a very intense operation with a for loop which is potentially quite slow, have you checked the performance of this plugin?

ahiguera-mx replied:
Here are a few numbers on Dali
for an  average background run `CPU times: user 1min 15s, sys: 15.6 s, total: 1min 31s` 
for an average Rd run `CPU times: user 4min 59s, sys: 57.6 s, total: 5min 57s`
for an average Kr run `CPU times: user 8min 22s, sys: 1min 36s, total: 9min 59s`
which includes loading peaks

JoranAngevaare replied:
comparing to run 12236, this seems much longer than any other plugin we had at that time I wrote this notebook:
https://github.com/XENONnT/workshop_zoom_2021_04_19/blob/master/1_xenonnt_data/1_Data_Structure_and_processing_tutorial.ipynb
![afbeelding](https://user-images.githubusercontent.com/22295914/155678742-bac19301-4b5a-412e-940a-2fc1d78a9863.png)

This being said, the times may differ a lot from run to run, so I don't know if the max ~350 s (total time) compares to the 600s wall-time you report here.

My suspicion is that this for-loop is probably a bottle neck. I  personally very much like [line-profiling ](https://pypi.org/project/line-profiler/) a lot in order to find bottlenecks and focus optimization on the parts of the code that are actually worth optimizing.

Let me know if you need some examples of how I would try and optimize for performance. A few general hints would concerning this block of code would be:
 - Figure out which part of the code takes the longest, if you spend a lot of time speeding something up that wasn't slow, you don't gain much (only because I've done this often myself, and it's a very tempting pitfall 😉 )
 - avoid for loops (unless you can do it in a numbafied function) or the numer of iterations is small
 - avoid `np.append`, every time you call that, you instruct numpy to allocate an entirely new block of memory which you then have to fill with the old stuff array

I can write you an example if you want, probably also useful for other people if @WenzDaniel or me are asking again and again if something is "fast" or "slow"

ahiguera-mx replied:
Hi @JoranAngevaare 
Good point, at the moment this plugin is at the level of `event_pattern_fit` in terms of slowness. We already have an idea on how to remove `np.append` for the reason you mentioned, but we would need to retrain. 
If the priority is to have this ready so analyst can use it, I think we can go ahead and meanwhile we work on the optimization. Also I would like to get other people opinion @tunnell @tinapeters 

Btw, once I tried to run sneakviz  to check this but I end up lost looking at the profile because all straxen calls , do you have an example for dummies on how to do this for a single straxen plugin

ahiguera-mx replied:
When running this plugin ~20% of the time is just loading peaks, I will try to debug more and see if can speed up 
```
Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    17                                           @profile
    18                                           def _compute_bayes(st):
    19                                           
    20         1      45984.0  45984.0      0.0      p = st.get_single_plugin('031711', 'peak_classification_bayes')
    21                                           
    22         1   21889439.0 21889439.0     21.0      peaks = st.get_array('031711', 'peaks')
    23                                               
    24         1   82340242.0 82340242.0     79.0      result =p.compute(peaks)

```

JoranAngevaare commented on straxen/plugins/event_w_bayes_class.py:
```diff
@@ -0,0 +1,55 @@
+        for peaks, name in zip([s1, s2, s1_alt, s2_alt], ['s1', 's2', 'alt_s1', 'alt_s2']):
+            peaks['time'] = events[f'{name}_time']
+            peaks['endtime'] = events[f'{name}_endtime']
+            fci = strax.fully_contained_in(peaklets, peaks)
```
If I remember correctly, you can't use this function as peaklets aren't sorted by time since an S1 can happen during an S2

ahiguera-mx replied:
n/a

JoranAngevaare commented on straxen/plugins/event_w_bayes_class.py:
```diff
@@ -0,0 +1,55 @@
+            mask = fci >= 0
+            fci = fci[mask]
+            result[f'{name}_s1_ln_prob'][fci] = peaklets[mask]['s1_ln_prob']
+            result[f'{name}_s2_ln_prob'][fci] = peaklets[mask]['s2_ln_prob']
```
I'm confused here, multiple peaklets an constitute a single peak. I don't understand how this is supposed to work

ahiguera-mx replied:
n/a we switch to peaks

JoranAngevaare commented on straxen/plugins/event_w_bayes_class.py:
```diff
@@ -0,0 +1,55 @@
+            result['time'] = events['time']
+            result['endtime'] = events['endtime']
+
+        # events can be made out of one peak, this is to enusure user should not look at prob
+        for name in (['s1', 'alt_s1', 'alt_s2']):
+            no_s_peak = np.where(events[f'{name}_index'] == -1)
+            if no_s_peak:
+                result[f'{name}_s1_ln_prob'][no_s_peak] = np.nan
+                result[f'{name}_s2_ln_prob'][no_s_peak] = np.nan
```
might be better to set everything to nan beforehand (top of this function) so you don't have to do an extra search where things are filled

ahiguera-mx replied:
I tried but since results has a predefined type 
`result = np.empty(len(events), dtype=self.dtype)`
if I set to nan I got `ValueError: cannot convert float NaN to integer` 
any suggestion here?

JoranAngevaare replied:
Ah, no so what I meant was setting this all the way at the top of this `compute` method:

```
for name in (['s1', 'alt_s1', 'alt_s2']):
    result[f'{name}_s1_ln_prob'] = np.nan
    result[f'{name}_s2_ln_prob'] = np.nan
``` 

The ones where you will find a peak for will get filled by your logic.

JoranAngevaare commented on straxen/plugins/event_w_bayes_class.py:
```diff
@@ -0,0 +1,55 @@
+
+        return dtype
+
+    def compute(self, peaklets, events):
```
If you were to do this on peaks (see earlier comment), it would be much faster since you could just select peaks on their start and endtime

ahiguera-mx replied:
done

JoranAngevaare reviewed (commented):
Thanks guys, the code looks carefully written and nicely documented etc.

Some things confuse me a bit as I will explain below. Additionally, I am wondering about the performance of this plugin. Did you check if there are any bottlenecks?

ahiguera-mx commented:
Thanks @JoranAngevaare for very useful comments, I've addressed.

JoranAngevaare commented on straxen/plugins/event_w_bayes_class.py:
```diff
@@ -0,0 +1,48 @@
+        result = np.empty(len(events), dtype=self.dtype) 
+
+        for name in ['s1', 's2', 'alt_s1', 'alt_s2']:
+            container = np.zeros(len(events), strax.time_fields)
+            container['time'] = events[f'{name}_time']
+            container['endtime'] = events[f'{name}_endtime']
+
+            mask = strax.touching_windows(peaks, container)
```
```suggestion
        # Select peaks based on their start time
        mask = np.in1d(peaks['time'], events[f'{name}_time'])
```

JoranAngevaare commented on straxen/plugins/event_w_bayes_class.py:
```diff
@@ -0,0 +1,48 @@
+
+            mask = strax.touching_windows(peaks, container)
+            result[f'{name}_ln_prob_s1'] = peaks['ln_prob_s1'][mask[:,0]]
+            result[f'{name}_ln_prob_s2'] = peaks['ln_prob_s2'][mask[:,0]]
+            result['time'] =  peaks[mask[:,0]]['time']
+            result['endtime'] =  peaks[mask[:,0]]['endtime']
```
```suggestion
            result[f'{name}_ln_prob_s2'] = peaks['ln_prob_s2'][mask[:,0]]
        result['time'] =  events['time']
        result['endtime'] =  events['endtime']
```

tunnell commented:
@JoranAngevaare will the speed things block you merging it?  Tina will work on it, but we really need to get it merged if people are going to test it more broadly.  Can we parallelize the work such that analysts can test this while Tina looks at the speed such that we don't make some last minute optimization mistake?

Also, what blocks the merge at this point?  Discussion hard to follow.  Do we know when the next release is?

l-althueser commented:
One thing I just noticed and missed before. We added the config files to CMT and latest global. We might want to consider using it here, right? If that's the case the default should be a cmt option to avoid confusion. If not then we should make it clear once more for me :)

ahiguera-mx commented:
> One thing I just noticed and missed before. We added the config files to CMT and latest global. We might want to consider using it here, right? If that's the case the default should be a cmt option to avoid confusion. If not then we should make it clear once more for me :)

Done, I change the config option to CMT, once is ready to merge I will add Bayes model to ONLINE

WenzDaniel commented:
> Can we parallelize the work such that analysts can test this while Tina looks at the speed such that we don't make some last minute optimization mistake?

Not a fan of it, but in the end it wont be a hard requirement, I think. The problem is, that we heard this promise already several times from different people, and in the end there are always the same 3-4 for people making the work and required changes. So we made a lot of bad experience here. 

> Also, what blocks the merge at this point? Discussion hard to follow. Do we know when the next release is?

I think the first requirement is that all tests pass which is not the case at the moment. Then I still have to review the PR. So far I have not had any time yet. At the moment we are preparing a release in which wont include this PR yet. There is another patch release planned by the end of this week, but I am not sure if this is a realistic time line. My feeling it will be mid next week, We are currently, fighting with CMT and the implementation of one of our new corrections. I spent yesterday the entire afternoon and evening to make it work in straxen.

WenzDaniel commented on straxen/plugins/bayes_peaklet_classification.py:
```diff
@@ -0,0 +1,138 @@
+    Returns the ln probability of a each event belonging to the S1 and S2 class.
+    Uses conditional probabilities and data parameterization learned from wfsim data.
+
+    :param peaks: peaks
+    :param waveforms: peaks waveforms in PE/ns
+    :param quantiles: quantiles in ns
+    :returns: the ln probability of a each peaklet belonging to S1 and S2 class
```
This looks a bit out of place as your class does not take those arguments.

WenzDaniel commented on straxen/plugins/bayes_peaklet_classification.py:
```diff
@@ -0,0 +1,138 @@
+    def setup(self):
+
+        self.class_prior = np.ones(self.n_bayes_classes)/self.n_bayes_classes
+        self.bins = self.bayes_config_file['bins']
+        self.cpt = self.bayes_config_file['cprob']
```
A comment about those would be good.

ahiguera-mx replied:
those parameters are explained in the URL config help

WenzDaniel commented on straxen/plugins/bayes_peaklet_classification.py:
```diff
@@ -0,0 +1,138 @@
+    Bayes Peaklet classification
+    Returns the ln probability of a each event belonging to the S1 and S2 class.
+    Uses conditional probabilities and data parameterization learned from wfsim data.
+
```
Can you maybe add also the reference to a note here since this is something new.

WenzDaniel commented on straxen/plugins/bayes_peaklet_classification.py:
```diff
@@ -0,0 +1,138 @@
+                    ln_prob_s2=ln_prob_s2
+                    )
+
+def compute_wf_and_quantiles(peaks: np.ndarray, bayes_n_nodes: int):
```
I think this function can be numbafied easily. Within your for-loop your can use a fixed length buffer instead of np.append as the length of your buffer is determined by the number of quartiles  and peaks is not it? The other thing which your have to change is `np.linspace(0., 1., bayes_n_nodes, endpoint=False)` into `np.linspace(0., 1., bayes_n_nodes)[:-1]` (assuming this is what endpoint false does).

ahiguera-mx replied:
I did an attempt to get rid of the np.append following your suggestion 
```
# create buffer to avoid np.append
frac_of_cumsum = np.empty(num_samples+1)
frac_of_cumsum[0] = 0.0
frac_of_cumsum[1:] = np.cumsum(data[i, :]) / np.sum(data[i, :])
# create buffer to avoid np.append
cumsum_steps = np.empty(bayes_n_nodes+1)
cumsum_steps[0:bayes_n_nodes] = np.interp(np.linspace(0., 1., bayes_n_nodes, endpoint=False), frac_of_cumsum, sample_number)
 cumsum_steps[-1:] = sample_number[-1]
quantiles[i, :] = cumsum_steps[1:] - cumsum_steps[:-1]
```

Locally passed all tests, however did not pass test on GitHub

WenzDaniel replied:
You mean with njit decorator? I think it cannot be numbafied due to `, endpoint=False` in linespace. You can check numbas very excellent [reference manual](https://numba.readthedocs.io/en/stable/reference/numpysupported.html).

ahiguera-mx replied:
My first attempt was to remove the np.append(), once that is working I can try to numbafied

WenzDaniel commented on straxen/plugins/bayes_peaklet_classification.py:
```diff
@@ -0,0 +1,138 @@
+    steps = np.arange(0, num_samples+1, step_size)
+
+    data = peaks['data'].copy()
+    dts = peaks['dt'].copy()
```
This line is not necessary is it?

ahiguera-mx replied:
fixed

WenzDaniel commented on straxen/plugins/bayes_peaklet_classification.py:
```diff
@@ -0,0 +1,138 @@
+    dts = peaks['dt'].copy()
+    data[data < 0.0] = 0.0
+    for i, p in enumerate(peaks):
+        sample_number = np.arange(0, num_samples+1, 1)*p['dt']
+        frac_of_cumsum = np.append([0.0], np.cumsum(data[i, :]) / np.sum(data[i, :]))
```
Why do you need this leading zero?

WenzDaniel commented on straxen/plugins/bayes_peaklet_classification.py:
```diff
@@ -0,0 +1,138 @@
+        cumsum_steps = np.append(cumsum_steps, sample_number[-1])
+        quantiles[i, :] = cumsum_steps[1:] - cumsum_steps[:-1]
+
+    for j in range(bayes_n_nodes):
+        waveforms[:, j] = np.sum(data[:, steps[j]:steps[j+1]], axis=1)
+    waveforms = waveforms/(dts*step_size)[:, np.newaxis]
```
Can you explain and maybe add a comment what you are doing here. It is not immediately obvious to me.

WenzDaniel commented on straxen/plugins/bayes_peaklet_classification.py:
```diff
@@ -0,0 +1,138 @@
+                      waveforms: np.ndarray, quantiles: np.ndarray):
+    """
+    Bin the waveforms and quantiles according to Bayes bins and compute inference
+    :param bins: Bayes bins
+    :param bayes_n_nodes: number of nodes or atributes
+    :param cpt: conditioanl probability tables
+    :param n_bayes_classes: number of classes
+    :param class_prior: class_prior
+    :param waveforms: waveforms
+    :param quantiles: quantiles
+    :return: ln probability per class S1/S2
```
Can you please add more information. Explaining e.g. a `car` with the word `car` is not really helpful.

WenzDaniel commented on straxen/plugins/bayes_peaklet_classification.py:
```diff
@@ -0,0 +1,138 @@
+    Compute waveforms and quantiles for a given number of nodes(atributes)
+    :param peaks:
+    :param bayes_n_nodes: number of nodes or atributes
+    :return: waveforms and quantiles
```
The waveforms are the down sampled versions of the original waveform right?

ahiguera-mx replied:
yes

WenzDaniel commented on straxen/plugins/bayes_peaklet_classification.py:
```diff
@@ -0,0 +1,138 @@
+    quantile_bin_edges = bins[1, :][bins[1, :] > -1]
+    quantile_num_bin_edges = len(quantile_bin_edges)
+
+    waveform_values = np.digitize(waveforms, bins=waveform_bin_edges)-1
+    waveform_values[waveform_values < 0] = int(0)
+    waveform_values[waveform_values > int(waveform_num_bin_edges - 2)] = int(waveform_num_bin_edges - 2)
+
+    quantile_values = np.digitize(quantiles, bins=quantile_bin_edges)-1
+    quantile_values[quantile_values < 0] = int(0)
+    quantile_values[quantile_values > int(quantile_num_bin_edges - 2)] = int(quantile_num_bin_edges - 2)
```
Can you explain what you are doing here I cannot follow. Also having the suffix `_values` is a bit miss leading. If I understand np.digitize correctly it will return indices not the waveform values.

WenzDaniel commented on straxen/plugins/event_info.py:
```diff
@@ -23,11 +23,12 @@ class EventInfo(strax.MergeOnlyPlugin):
                   'corrected_areas',
                   'energy_estimates',
+                  'event_w_bayes_class',
                   # 'event_pattern_fit', <- this will be added soon
                   ]
     save_when = strax.SaveWhen.ALWAYS
     provides = 'event_info'
-    __version__ = '0.0.2'
+    __version__ = '0.0.3'
```
Can we exclude it from event_info like we do for `event_pattern_fit`. Then we can process it fairly independently from the main data and hence performance is less of an issue if things arrive late.

ahiguera-mx replied:
I will leave this up to you

WenzDaniel commented on straxen/plugins/event_w_bayes_class.py:
```diff
@@ -0,0 +1,42 @@
+            result[f'{name}_ln_prob_s1'] = np.nan
+            result[f'{name}_ln_prob_s2'] = np.nan
+            # Select peaks based on their start time
+            mask = np.in1d(peaks['time'], events[f'{name}_time'])
+            mask_ev = np.in1d(events[f'{name}_time'], peaks['time'])
+            result[f'{name}_ln_prob_s1'][mask_ev] = peaks['ln_prob_s1'][mask]
+            result[f'{name}_ln_prob_s2'][mask_ev] = peaks['ln_prob_s2'][mask]
```
You could use also our strax-methods like `strax.fully_contained_in` and `s1_index` etc. Not sure if it will be faster,

ahiguera-mx replied:
`strax.fully_contained_in` is not an option since and S1 can be contained within an S2

WenzDaniel commented on tests/test_bayes_plugin.py:
```diff
@@ -0,0 +1,37 @@
+        self.st.make(self.run_id, 'peaks')
+        self.st.make(self.run_id, self.target)
+        bayes = self.st.get_array(self.run_id, self.target)
+        prob = np.where(bayes['ln_prob_s1'] > 0.0)
+        assert bayes[prob].size == 0, "ln prob gratter than zero, impossible"
```
`assert np.any(bayes['ln_prob_s1'] > 0)` does the same and is shorter.

ahiguera-mx replied:
done

WenzDaniel reviewed (commented):
I think it looks mostly good. Here some comments.

ahiguera-mx commented:
move this to https://github.com/XENONnT/straxen/pull/954
